ttnn/cpp/ttnn/tensor/tensor_impl.hpp

```diff
@@ -142,6 +142,13 @@ inline std::vector<T> convert_layout_tile_to_row_major(
         transpose_of_faces);
 }
 
+template <typename T>
+std::vector<T> convert_logical_data_to_physical_data(const std::vector<T>& logical_data, const TensorSpec& tensor_spec);
+
+template <typename T>
+std::vector<T> convert_physical_data_to_logical_data(
+    const std::vector<T>& physical_data, const TensorSpec& tensor_spec);
+
```

Comment:
Great utils! Can you clarify in a comment here what exactly is happening under the hood? +1 to your other comment regarding physical vs logical data. But it is still very useful to mention the expectations: conversion to tilized layout (and to row major layout on the way back) + zero value padding.

Reply:
Can we also use `tt::stl::Span` here, and perhaps add dedicated C++ unit tests here? There is quite a lot of logic.

Reply:
Btw, what happens if we want to create a tilized tensor but the dims aren't divisible by the tile dimensions. Are we going to pad automatically, or is it on the user? Does this PR change anything w.r.t. that?

Reply:
There are dedicated C++ unit tests already in `test_sharding_with_alignment.cpp`. All the logic here is moved from there. It does everything from row major host data <-> physical data described by tensor spec, so yet it adds padding if needed.

I agree about the use of vector/buffer vs. tensor. I think all data owned by ttnn tensors needs to be accompanied by tensor spec in general. If it means that all these functions should operate with tensors, we should do a larger refactor specifically for that.